### PR TITLE
Added redux, react-router, neat, google analytics

### LINKED ISSUE #12: React app setup
- [ ] web pack
       dev & production
- [ ] ES6
- [ ] react
- [ ] mocha

Use teachers portal as a guide

### LINKED ISSUE #45: Google Analytics
- [X] Create project in ODL GA account and -ci and -release profiles. 
        CI tag is UA-5145472-14
        Release tag (already in use) is UA-5145472-10
- [x] Invite all the necessary people to the GA account
- [x] add UA key to config
- [x] ? React coding?

### PR DESCRIPTION
**Pull request by jamiefolsom**
#### What are the relevant tickets?

Fixes #12 
Fixes #45

#### What's this PR do?

Adds some front-end packages, including redux, react-router, google analytics, bourbon neat.  

#### Where should the reviewer start?

package.json

#### How should this be manually tested?

run docker-compose up, there should be no console errors. Visit /dashboard, and you should see a checkbox, which when clicked, shows redux actions in the js console.

#### Any background context you want to provide?

Setting up the infrastructure for the react app, based on what we learned in Teachers' Portal.

#### Screenshots (if appropriate)

N/A

### CODE REVIEW

File: static/js/inner.js [hunk 1]
```diff
@@ -1,3 +1,30 @@
...
+import ReactDOM from 'react-dom';
+import App from './containers/App';
+import '../sass/layout.scss';
+import { Provider } from 'react-redux';
+import { Router, Route, IndexRoute } from 'react-router';
+import { DevTools, DebugPanel, LogMonitor } from 'redux-devtools';
+import ga from 'react-ga';
+
+let debug = SETTINGS.reactGaDebug === "true";
+ga.initialize("abcdefg1234567", { debug: debug });
```
Comment (noisecapella) [inline static/js/inner.js]: Can you change this to use `SETTINGS.gaTrackingID`?

File: ui/templates/index.html [hunk 2]
```diff
@@ -5,5 +5,6 @@
 
 {% block content %}
 <h1>{% trans "Hi, I'm micromasters" %}</h1>
+<div id="container"></div>
```
Comment (noisecapella) [inline ui/templates/index.html]: I don't think we're using React at all for this page since it's not part of the dashboard

Comment (jamiefolsom) [inline ui/templates/index.html]: Ah, ok.

File: static/js/inner.js (hunk 1 above)
Comment (jamiefolsom) [inline static/js/inner.js]: Done.

File: static/js/actions/index_page.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { createAction } from 'redux-actions';
+import { sendGoogleAnalyticsEvent, sendGoogleEcommerceTransaction } from '../util/util';
```
Comment (justinabrahms) [inline static/js/actions/index_page.js]: Unused?

File: static/js/util/util.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    label: label,
+  };
+  if (value !== undefined) {
+    event.value = value;
+  }
+  ga.event(event);
+}
+
+export function sendGoogleEcommerceTransaction(id, revenue) {
+  ga.plugin.require('ecommerce');
```
Comment (justinabrahms) [inline static/js/util/util.js]: We won't have any ecommerce stuff on this app, afaik.

File: ui/views.py [hunk 5]
```diff
@@ -25,7 +26,17 @@ def index(request):
...
-    return render(request, "index.html", context={
-        "style_src": get_bundle_url(request, "style.js"),
-        "outer_src": get_bundle_url(request, "outer.js"),
-    })
+    js_settings = {
+        "gaTrackingID": settings.GA_TRACKING_ID,
+        "reactGaDebug": settings.REACT_GA_DEBUG
+    }
+
+    return render(
```
Comment (justinabrahms) [inline ui/views.py]: I think this is the view for the homepage, not the "react app"?

Comment (jamiefolsom) [inline ui/views.py]: Right you are. I added a view for "dashboard".

File: static/js/actions/index_page.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import { sendGoogleAnalyticsEvent } from '../util/util';
+
+export function event() {
```
Comment (justinabrahms) [inline static/js/actions/index_page.js]: Would suggest renaming this to something that is a bit more descriptive like `emitPageview` or similar.

Comment (justinabrahms) [inline static/js/actions/index_page.js]: Actually, I don't think this is used anywhere.

File: static/js/store/configureStore.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    applyMiddleware(
+      thunkMiddleware
+    )
+  )(createStore);
+}
+
+export default function configureStore(initialState) {
+  const store = createStoreWithMiddleware(rootReducer, initialState);
+
+  if (module.hot) {
```
Comment (justinabrahms) [inline static/js/store/configureStore.js]: Not a blocker. Do you know what this does? I'd guess it's like checking `globals()['hot']` in python, but I don't see where that would get added to the current state by looking at this file.

File: ui/urls.py [hunk 8]
```diff
@@ -3,8 +3,9 @@
 """
 from django.conf.urls import url
 
-from ui.views import index
+from ui.views import index, dashboard
 
 urlpatterns = [
-    url(r'^$', index, name='ui-index'),
+url(r'^dashboard/', dashboard, name='ui-dashboard'),
```
Comment (justinabrahms) [inline ui/urls.py]: You lost your indentation here.

File: static/js/store/configureStore.js (hunk 7 above)
Comment (jamiefolsom) [inline static/js/store/configureStore.js]: No, actually -- this is copy/pasted from teachersportal.

**Comment by justinabrahms** [comment]
Looks like this could also use a rebase to account for Anna's homepage work. :)

File: ui/urls.py (hunk 8 above)
Comment (jamiefolsom) [inline ui/urls.py]: Good catch. Fixed.

**Comment by jamiefolsom** [comment]
@justinabrahms rebased.

**Comment by giocalitri** [comment]
@jamiefolsom there are still some problems

**Comment by jamiefolsom** [comment]
@giocalitri @justinabrahms with @noisecapella, I think we got most of the remaining issues addressed. Now dealing with some lint issues; will push again shortly.

**Comment by jamiefolsom** [comment]
@noisecapella @giocalitri @justinabrahms ready for review. Thanks.

**Comment by noisecapella** [comment]
I'll push to fix the pylint issue

File: static/js/store/configureStore.js (hunk 7 above)
Comment (noisecapella) [inline static/js/store/configureStore.js]: I originally added it to teachersportal, from here: https://webpack.github.io/docs/hot-module-replacement.html

File: ui/templates/base.html [hunk 9]
```diff
@@ -10,7 +13,7 @@
   <body>
     {% block content %}
     {% endblock %}
-  <script type="text/javascript" src="{{ outer_src }}"></script>
+  <script type="text/javascript" src="{{ inner_src }}"></script>
```
Comment (noisecapella) [inline ui/templates/base.html]: Can you revert this and add this line to dashboard.html?

**Comment by noisecapella** [comment]
Just the one comment, other than that I think it looks good to merge :+1:

File: ui/templates/base.html (hunk 9 above)
Comment (jamiefolsom) [inline ui/templates/base.html]: OK, I've moved `outer_src` to `index.html` and `inner_src` to `dashboard.html`. Make sense @noisecapella? Thanks.

**Comment by jamiefolsom** [comment]
@noisecapella let me know if that adjustment to the script source links looks good to you, and I will merge if so.

**Comment by noisecapella** [comment]
:+1:

**Pull request merged**